internal/operator/reconciliations/api_gateway/reconciliation.go

```diff
@@ -0,0 +1,149 @@
+package api_gateway
+
+import (
+	"context"
+	"fmt"
+	"github.com/kyma-project/api-gateway/controllers"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/util/retry"
+
+	operatorv1alpha1 "github.com/kyma-project/api-gateway/apis/operator/v1alpha1"
+	"github.com/kyma-project/api-gateway/internal/operator/resources"
+	networkingv1alpha3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type ApiGatewayReconciliation interface {
+	Reconcile(ctx context.Context, apiGatewayCR operatorv1alpha1.APIGateway, apiGatewayResourceListPath string) (operatorv1alpha1.APIGateway, controllers.Status)
+}
+
+type Reconciliation struct {
+	Client client.Client
+}
+
+const (
+	reconciliationFinalizer string = "apigateways.operator.kyma-project.io/api-gateway-reconciliation"
+	dafaultGatewayName      string = "kyma-gateway"
+	dafaultGatewayNS        string = "kyma-system"
+)
+
+var checkDefaultGatewayReference = func(ctx context.Context, c client.Client, res resources.Resource) bool {
```

Comment:
I think this logic should also be moved to the gateway package, since this is tightly coupled to it. Also we can re-use this logic here to block the deletion of the Kyma Gateway, because right now it [blocks on any APIRule](https://github.com/kyma-project/api-gateway/blob/mod-dev/internal/gateway/kyma_gateway.go#L30). Having different criteria doesn't make sense.

Reply:
Will move it to gateway package, but I am unsure about the logic - https://github.com/kyma-project/api-gateway/issues/559 states it should check if APIRule is referring the gateway. On the mentioned code, it blocks on any APIRule. If I am remembering correctly first we agreed not to check if APIRule refers the gateway, but thought this changed. Should we sync on this again?

Reply:
Decided on that, we move the condition for referring APIRules and VSs to gateway deletion and check for existing any APIRules to API-Gateway CR deletion conditions.